src/test/scala/ai/privado/languageEngine/java/tagger/source/JavaSourceTaggingTest.scala

```diff
@@ -0,0 +1,117 @@
+
+    "tag a firstName identifier" in {
+      val identifierNodes = cpg.identifier("firstName").tag.nameExact(Constants.id).l
+      identifierNodes.size shouldBe 1
+      identifierNodes.value.head shouldBe "Data.Sensitive.FirstName"
+    }
+
+    "tag fieldAccess of firstName" in {
+      // Note - this test is Fails with the current query on cpg.method.callIn, but works fine on cpg.call
+      val identifierNodes = cpg.fieldAccess.tag.nameExact(Constants.id).l
+      identifierNodes.size shouldBe 2
+      identifierNodes.value.head shouldBe "Data.Sensitive.FirstName"
+    }
+  }
+
+  "Simple use case of Member variables of class matching with source rules" should {
+    val cpg = code("""
+        |public class User {
+        |   public String firstName;
+        |
+        |   public String getFirstName() {return firstName;}
+        |   public void setFirstName(String firstName) {this.firstName = firstName;}
+        |}
+        |""".stripMargin)
+      .moreCode("""
+        |public class Test {
+        |   public static void main(String[] args){
+        |       User us = new User();
+        |       System.out.println(us);
+        |   }
+        |}
+        |""".stripMargin)
+      .generateScanResult()
+
+    "tag a us identifier" in {
+      val identifierNodes = cpg.identifier.where(_.tag.nameExact("catLevelOne").valueExact("DerivedSources")).l
+      // TODO: Ideally it should be 2 nodes at line no 4 and line 5. However Line no 4 Identifier node is getting tagged two times.
+      // As it doesn't harm in processing, we will keep this issue on back burner.
+      identifierNodes.size shouldBe 3
+      identifierNodes.code.l shouldBe List("us", "us", "us")
+    }
+  }
+
+  "Templated use case of Derived source" should {
+    val cpg = code("""
+        |public class User {
+        |   public String firstName;
+        |
+        |   public String getFirstName() {return firstName;}
+        |   public void setFirstName(String firstName) {this.firstName = firstName;}
+        |}
+        |""".stripMargin)
+      .moreCode("""
+        |import java.util.Optional;
+        |public class Test {
+        |   public Optional<User> findUser() {
+        |     User userone = new User();
+        |     return Optional.ofNullable(userone);
+        |   }
+        |   public static void main(String[] args){
+        |       Optional<User> userOption = findUser();
+        |       User user = userOption.orElseThrow(() ->
+        |            new RunTimeExceptionPlaceHolder("User doesn't exist!!")
+        |       );
+        |       System.out.println(user);
+        |   }
+        |}
+        |""".stripMargin)
+      .generateScanResult()
+
+    "tag a userone and user identifier" in {
+      val identifierNodes = cpg.identifier.where(_.tag.nameExact("catLevelOne").valueExact("DerivedSources")).l
+      identifierNodes.size shouldBe 5
+      identifierNodes.code.l shouldBe List("userone", "userone", "userone", "user", "user")
```

Comment:
One suggestion while writing test case is to also include some metadata like linenumber, which can help in relating in  what all it getting tag, as currently just by looking at 5 identifiers and there code, is not self explainatory

Reply:
make sense

Reply:
done